### PR TITLE
Implement generic overrides

### LINKED ISSUE #400: Generic overrides
As per a [comment](https://github.com/kubernetes-sigs/federation-v2/issues/367#issuecomment-433912857) on #367, the recent implementation of `kubefed2 federate` has highlighted the need to support generic overrides (in which _almost_ [1] any field can be overridden).  Without generic overrides, generated override types could vary by deployment and this would complicate productization.

A simple path-based proposal:

```yaml
apiVersion: generated.federation.k8s.io/v1alpha1
kind: FederatedServiceOverride
metadata:
  name: foo
  namespace: bar
spec:
  overrides:
  - clusterName: cluster1
     clusterOverrides:
     - path: spec.type
       value: NodePort
  - clusterName: cluster2
     clusterOverrides:
     - path: spec.type
       value: LoadBalancer
```

This simple approach does not preclude the future implementation of optional json-patch based overrides:

```yaml
apiVersion: generated.federation.k8s.io/v1alpha1
kind: FederatedServiceOverride
metadata:
  name: foo
  namespace: bar
spec:
  overrides:
  - clusterName: cluster1
     clusterOverrides:
     - path: spec.type
       value: NodePort
  - clusterName: cluster2
     clusterOverrides:
     # Equivalent json patch directive
     - json: { "op": "replace", "path": "/spec/type", "value": "NodePort" }
```

1: `name` and `namespace` will not be allowed to be overridden to ensure associativity is maintained between primitive resources and resources propagated to member clusters.

### PR DESCRIPTION
**Pull request by marun**
Generic overrides removes the requirement to define fields to override in the FederatedTypeConfig and ensures that deployments of federation won't differ in the fields they allow to be overridden for a given federated type.

Fixes #400

### CODE REVIEW

**Comment by marun** [comment]
I'm in the process of rewriting the override update code for replica scheduling to resolve the CI failure.

**Comment by shashidharatd** [comment]
This is neat. A much needed feature. LGTM. Thanks @marun !

**Comment by marun** [comment]
I've added a couple more commits to cleanup reading and writing overrides to the unstructured object.

**Comment by marun** [comment]
And one more enabling (and testing) array overrides.

**Comment by marun** [comment]
Rebased and updated helm chart crds.

**Comment by marun** [comment]
@danehans I was hoping for a more substantive review before merge.  Please avoid the use of `/lgtm` for non-trivial changes since we can't easily disable self-approval (by the submitter) at this time.

**Comment by marun** [comment]
@mvazquezc I'm assuming you're asking if it's possible to set the value of an array item.  The path-based mechanism implemented in this PR isn't able to dereference array indexes, and the json-path based scheme mentioned in #400 would be required to do so.  However, it should be possible to work around this by federating a config map and varying its values by cluster, and then consuming those config map values as environment variables.

**Comment by mvazquezc** [comment]
> @mvazquezc I'm assuming you're asking if it's possible to set the value of an array item. The path-based mechanism implemented in this PR isn't able to dereference array indexes, and the json-path based scheme mentioned in #400 would be required to do so. However, it should be possible to work around this by federating a config map and varying its values by cluster, and then consuming those config map values as environment variables.

Okay, so let's wait for the json-path feature, meanwhile I can use the configmap approach.

Thanks for the clarification.

File: pkg/controller/util/overrides.go [hunk 1]
```diff
@@ -19,59 +19,123 @@ package util
...
 	if !ok {
-		return nil, fmt.Errorf("Missing spec.overrides for %s %q: %v", overrideKind, qualifiedName, err)
+		return nil, fmt.Errorf("%s not found", overridesPath)
 	}
 
-	overridePaths := typeConfig.GetOverridePaths()
-	if len(overridePaths) == 0 {
-		return nil, fmt.Errorf("Override paths are missing for %q", typeConfig.GetTarget().Kind)
-	}
+	for i, overrideInterface := range rawOverrides {
```
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controller/util/overrides.go]: Thank you for the review! 

I think a panic is theoretically possible if the input object was arbitrary in structure, but the combination of crd validation and the structure of `Unstructured` make it unlikely if not impossible in practice.  That said, I think json marshaling to a concrete type makes more sense than interacting directly with an unstructured type, which would remove the need for type assertions.

**Pull request merged**